src/rust/aggregator/src/pipelines/order_history_pipelines.rs

```diff
@@ -0,0 +1,598 @@
+                .into_iter()
+                .map(|e| (e.market_id, e.group_id));
+        let market_id_to_group = HashMap::from_iter(market_id_to_group);
+
+        let last_indexed_timestamp = sqlx::query_file!(
+            "sqlx_queries/order_history_pipelines/get_last_indexed_timestamp.sql"
+        )
+        .fetch_optional(&mut transaction as &mut PgConnection)
+        .await
+        .map_err(to_pipeline_error)?
+        .map(|e| e.time);
+        // This can happen if no limit orders have been posted yet.
+        let last_indexed_timestamp = if let Some(last_indexed_timestamp) = last_indexed_timestamp {
+            last_indexed_timestamp
+        } else {
+            return Ok(());
+        };
+
+        let latest_event_timestamp = sqlx::query_file!(
+            "sqlx_queries/order_history_pipelines/get_latest_event_timestamp.sql"
+        )
+        .fetch_one(&mut transaction as &mut PgConnection)
+        .await
+        .map_err(to_pipeline_error)?
+        .time;
+        let latest_event_timestamp = if let Some(latest_event_timestamp) = latest_event_timestamp {
+            latest_event_timestamp
+        } else {
+            return Ok(());
+        };
+
+        let timestamps = generate_timestamps(last_indexed_timestamp, latest_event_timestamp);
+
+        for timestamp in &timestamps {
+            self.get_state_from_timestamp_with_state(timestamp, &mut transaction)
+                .await?;
+            if let Some(state) = &self.state {
+                transaction = calculate_spreads_and_liquidities(
+                    &address_to_group,
+                    &market_id_to_group,
+                    &mut self.last_prices,
+                    state,
+                    timestamp,
+                    transaction,
+                )
+                .await?;
+            }
+        }
+        {
+            if let Some(last) = timestamps.last() {
+                sqlx::query_file!(
+                    "sqlx_queries/order_history_pipelines/update_last_indexed_timestamp.sql",
+                    last,
+                )
+                .execute(&mut transaction as &mut PgConnection)
+                .await
+                .map_err(to_pipeline_error)?;
+            }
+        }
+        transaction.commit().await.map_err(to_pipeline_error)?;
+        Ok(())
+    }
+}
+
+async fn calculate_spreads_and_liquidities(
+    address_to_group: &HashMap<String, i32>,
+    market_id_to_group: &HashMap<BigDecimal, i32>,
+    last_prices: &mut HashMap<BigDecimal, Option<BigDecimal>>,
+    state: &SpreadsState,
+    timestamp: &DateTime<Utc>,
+    transaction: Transaction<'static, Postgres>,
+) -> Result<Transaction<'static, Postgres>, PipelineError> {
+    let tx = Arc::new(RwLock::new(transaction));
+    let (res1, res2) = tokio::join! {
+        liquidities(&address_to_group, &market_id_to_group, last_prices, state, timestamp, tx.clone()),
+        spreads(state, timestamp, tx.clone())
```

Comment:
Can we pass the transaction by reference, e.g. `&txn` or similar? I'm concerned that if this is passing a clone of the txn that we might have unexpected behavior

Reply:
We are putting it in a `RwLock` to avoid concurrent access to the transaction struct. Then, we are putting it in an `Arc` (an Atomic Reference Counting pointer). Then cloning the `Arc`, not the transaction or `RwLock`. This is done to run both the `liuidities` and `spreads` function concurrently. We cannot just pass `&mut txn`, as it would be double mutable borrowing, which is forbidden by the rust compiler.